Correlation length and full normalisation

calum-chamberlain opened the pull request:
Correlations for the last sample were systematically missed. This PR:
1. Increases the step end point for `arch = "precise"`,
2. Corrects the position used for accessing the appropriate `csum_square_data` for `arch = "cpu"`
3. Gets to the end of `n_samples_data` for `arch = "gpu"`
4. Implements a full normalization option.

I found these when working on getting my EQcorrscan FMF wrapper working fully - it might be worth having a test in the test-set that compares against a *really* naive implementation of correlation to ensure that the correlations are actually correct, not just that they compare favorably with each other?

I haven't done anything about matlab implementations.

---

Now included:
> p.s. Because I'm trying to get FMF working well within EQcorrscan (because it is so much faster than EQcorrscan correlations in most cases) you may see an additional PR that implements an option for full normalisation. I tried running FMF for my Kaikoura matching and ended up with a significantly lower MAD, and hence many more false detections than when using a fully normalised correlation.

calum-chamberlain commented:
I also noticed that stopping iterations at:
```c
stop_i = 1 + (n_samples_data - n_samples_template) - max_moveout);
```
discounts the last samples up to `max_moveout` from all channels that have moveouts less than max_moveout.  This results in differently defined correlation stacks at the start (where correlations from invdividual channels are allowed) to the end (where the correlation sum stops as soon as one channel is lost).

[This commit](https://github.com/beridel/fast_matched_filter/pull/38/commits/a868cb4a4a0505ac23bb19fdd4b2057fb647fe64) fixes this for the CPU, but I haven't made this change for the GPU yet.

calum-chamberlain commented:
I ended up implementing full normalisation here as well.  This isn't supported on `arch="cpu"` because of the moving sum of squares.

This now passes all EQcorrscan tests with normalisation="full".

calum-chamberlain commented:
Also, this definitely isn't optimized.  It would probably be faster to pre compute the moving mean.  I can add that in this week.

ebeauce commented:
Thank you Calum for all these great additions.

I have been through somewhat quickly, but I approve of most changes. I agree that making sure the correlation stacks at the beginning and at the end are defined consistently is best, but I actually would go for the most conservative approach of discarding any samples where all channels do not contribute. Doing otherwise could lead to unexpected results where the end (or the beginning) would produce lot's of detections when only one station (and a bad one, to take the worst case scenario) would make a bad quality CC time series.

I can see how to optimize the code next week and update this pull request. I also would be in favor of using another word than "naive" for the simplified CC definition, as the word "naive" has a negative connotation that could make users go for the full definition when it's not necessary, and thus running a slow version of FMF. I suggest "simple" or "short" instead.

calum-chamberlain commented:
Ah yes, apologies for the "naive" - happy to change that, and keen to leave it as the default to retain the same behavior in FMF.

RE defining correlations at beginning and end consistently: I'm in favor of the current approach for EQcorrscan, mostly because we cope with this externally (by overlapping the data chunks).  So if you do decide to go down the route of not having correlations at either end with only a few stations (which I agree is likely to lead to missed detections at the start and end of data chunks) then I would like to implement an option to retain the less-robust implementation here. Maybe this could be defined in the wrappers (Python and Matlab) rather than in the C and cuda source code?

Myself and Felix ran some timings over [here](https://github.com/eqcorrscan/EQcorrscan/pull/464) - a lot of the time spent calling FMF from EQcorrscan is "wasted" time spent reshaping templates (which is similar for all our correlation backends - eventually I will move EQcorrscan to more sensible data structures), so take the timings with a pinch of salt - the difference in times is mostly due to FMF though.

ebeauce commented:
Ok we can try to add another key word argument to tell FMF to return or not these less reliable CCs.

I see that you are running tests with quite long templates. I'm curious to see how FMF compares with EQcorrscan on these, as I think it's not an ideal case for making an optimal use of the GPUs (if the number of samples per template >> 512, one block of the GPU grid needs more time to store the data into shared memory). Keep in mind that you might into problems with the GPUs' shared memory, in which case you'll have to shorten the templates (or use more recent GPUs :)). Last point: the version of nvcc that was used to compile FMF is an important information when comparing speeds because later versions show significant speed ups with regard to earlier ones (up to 10-20%).

calum-chamberlain commented:
> I see that you are running tests with quite long templates. I'm curious to see how FMF compares with EQcorrscan on these, as I think it's not an ideal case for making an optimal use of the GPUs

Interesting point - I was (and am) running mine with shorter templates (4s @ 30 Hz) and everything seems to be going well. I hadn't picked up that Felix was using such long templates. Thanks for the tips!

> RE defining correlations at beginning and end consistently: I'm in favor of the current approach for EQcorrscan, 

I thought about this overnight and I think I am *a)* wrong, and *b)* think I prefer your way of doing this. I would be happy to revert to zeros at the end in that stack as you had it (e.g. reverting the change to [this line](https://github.com/beridel/fast_matched_filter/pull/38/files#diff-b5138fca9691acd324cf5cbaa26fe72c5fc650db2769875beea71e7f15d3134bR60) I think, but checking what that does to a single-channel as well).  

For some reason I thought that the start of the correlation sum also had reduced channel numbers, but I see that this is wrong. Everything is shifted "left" by it's moveout and any preceding samples trimmed from the stack. I guess ideally we could trim the zeros at the end as well, but that would mean different shaped correlation sums for different templates which would be hard to handle. I think retaining the zeros as you had, rather than returning the reduced quality correlations (as EQcorrscan does) is more robust, and I think I should change EQcorrscan to match this. Sorry for my confusion!

ebeauce commented on fast_matched_filter/src/matched_filter.c:
```diff
@@ -56,7 +57,7 @@ void matched_filter(float *templates, float *sum_square_templates, int *moveouts
 
         start_i = (int)(ceilf(abs(min_moveout) / (float)step)) * step;
-        stop_i = n_samples_data - n_samples_template - max_moveout;
+        stop_i = 1 + (n_samples_data - n_samples_template); // - max_moveout);
```
Revert this line to its initial form?

ebeauce commented on fast_matched_filter/src/matched_filter.c:
```diff
@@ -70,7 +71,7 @@ void matched_filter(float *templates, float *sum_square_templates, int *moveouts
                                                         n_samples_data,
                                                         n_stations,
-                                                        n_components);
+                                                        n_components, i);
```
This new extra argument is no longer necessary if we go for the conservative approach, right?

ebeauce commented on fast_matched_filter/src/matched_filter.c:
```diff
@@ -80,7 +81,8 @@ void matched_filter(float *templates, float *sum_square_templates, int *moveouts
                    float *data, double *csum_square_data, float *weights,
-                   int n_samples_template, int n_samples_data, int n_stations, int n_components) {
+                   int n_samples_template, int n_samples_data, int n_stations, int n_components, 
+                   int i) {
```
Same as previous comment, I believe the last argument isn't necessary.

ebeauce commented on fast_matched_filter/src/matched_filter.c:
```diff
@@ -95,6 +97,8 @@ float network_corr(float *templates, float *sum_square_template, int *moveouts,
             component_offset = station_offset + c;
             if (weights[component_offset] == 0) continue;
 
+            if ((i + moveouts[component_offset]) > n_samples_data - n_samples_template) continue;
```
This condition does not need to be tested if we already make sure we stop correlating when the first channel goes out of the data stream.

ebeauce commented:
I'm getting a bit lost in the commits. I commented the lines of the new code where I think we should revert the changes if we want to stick to the initial method for returning CCs at the end of the data.

calum-chamberlain commented:
I think I have reverted that all now. I also caught a niche bug in the GPU code where `cs` could end up being negative.  I found this for small `chunk_size` and relatively large `moveout`. I added a condition in there - it could be in a different spot though to reduce the number of times it is hit (e.g. by setting `NCHUNK` to a smaller value if `n_size_t / chunk_size > NCHUNK - 1`?

I also changed from `"naive"` to `"short"`.

ebeauce commented:
Good catch for the negative chunk_size! I always process ~ 1 day long data so I have never run into a negative cs, but that's totally something that can happen.
I have added one minor edit to the doc string of the matched_filter routine. Otherwise, I have reviewed all the commits and it looks good to me. Should we merge the branch as is, and release a 1.4.1 later if we find ways of optimizing the new normalization? (I am finishing my PhD in 2 months, so it might take me some time before I can take time to work on this)

ebeauce reviewed (approved):
Good to go, I think.

calum-chamberlain commented:
Sounds good to me - thanks for taking the time to help out with this @ebeauce especially at such a busy time! Good luck with finishing up!
